app/src/main/java/ch/epfl/sdp/item/Trap.java

```diff
@@ -0,0 +1,72 @@
+package ch.epfl.sdp.item;
+
+import ch.epfl.sdp.artificial_intelligence.Updatable;
+import ch.epfl.sdp.entity.EntityType;
+import ch.epfl.sdp.entity.Player;
+import ch.epfl.sdp.entity.PlayerManager;
+import ch.epfl.sdp.game.Game;
+import ch.epfl.sdp.map.Displayable;
+import ch.epfl.sdp.map.GeoPoint;
+
+/**
+ * A trap is an item a player can drop at his current localization and if another player walks by the trap, he will take damage
+ */
+public class Trap extends Item implements Updatable, Displayable {
+    private GeoPoint trapPosition;
+    private int damage;
+    private int radius;
+    private Player owner;
+
+    /**
+     * This is a constructor for traps
+     * @param damage The damage the trap will inflict on the other player that walks by
+     * @param radius The radius of the circle in which the other player needs to be in order for the trap to activate
+     */
+    public Trap(int damage, int radius) {
+        super(String.format("Trap (%d)", damage), String.format("Deal %d health points damages to the player that walks on it", damage));
+        this.damage = damage;
+        this.radius = radius;
+    }
+
+    @Override
+    public void use() {
+        //This is called by the player that has the item, so getUser should return the correct player
+        //The trick is that we need to save the reference to the player, since the update method will be called by the server and thus,
+        //getUser would return the wrong player (the server)
+        owner = PlayerManager.getUser();
+        trapPosition = owner.getLocation();
+        Game.addToUpdateList(this);
+        if (owner == PlayerManager.getUser()) {
+            Game.addToDisplayList(this);
+        }
+    }
+
+    @Override
+    public void update() {
+        boolean hasSomeoneTakenDamage = false;
+        for (Player p : PlayerManager.getPlayers()) {
+            if (p.getLocation().distanceTo(trapPosition) < radius && p != owner) {

```

Comment:
Instead of having "p!=owner", isn't it better to have !p.equals(owner) ? Because here, you're comparing the references and not the object itself. I'm not sure they have the same references and "p!=owner" may always bet true, even if "p" is the user. I'm not sure though

Reply:
Normally, comparing references here should be fine, as we just want to be sure it's the correct player. When comparing references, the problem is usually that you can have almost everytime a false output

Reply:
Alright